### PR TITLE
MessageList Virtualisation

### PR DESCRIPTION
**Pull request by mahboubii**
Using [React-Virtuoso](https://github.com/petyosi/react-virtuoso) to speed up the message list component

**Checklist**

- [x] Basic virtualization
- [x] Reverse infinite scroll
- [x] Scroll down on add new message
- [x] Stick to the bottom(follow new messages)
- [x] Loading UI
- [x] Render system events
- [x] More advanced Message UI
- [x] Message UI styling ~~and Grouping~~
- [x] Break down the component into separate files
- [x] Show connection status changes, better to have it as a standalone component next to the message list
- [x] Notification UI for new messages from other users
- [x] Basic Input component without attachments, ...
- [x] Add doc about how to properly use this and its shortcomings
- [x] Placeholder prop for fast scroll behavior
- [x] Show images in gallery mode
- [x] Message actions (delete/~~edit/flag/mute/ban~~)
- [ ] Possible attachments support with fixed size and placeholders
- [x] Add tests
- [ ] ~~Render date separators~~ not supported due to nondeterministic height

### CODE REVIEW

**Comment by tigershen23** [comment]
Hi friends! Customer who is eagerly awaiting this feature here :). Looks great so far. Will this component support a custom Message component? We have customized the Message component quite a bit and it would be great to plug and play. At a minimum, will it share CSS class names where appropriate?

**Comment by mahboubii** [comment]
> Hi friends! Customer who is eagerly awaiting this feature here :). Looks great so far. Will this component support a custom Message component? We have customized the Message component quite a bit and it would be great to plug and play. At a minimum, will it share CSS class names where appropriate?

Hey @tigershen23, it's most likely that we add support for custom Message component but keep in mind all virtualization libraries has some shortcomings, the main one in our case is you probably can't have a decent render performance if your Message component has dynamic heights.

**Comment by tigershen23** [comment]
Got it, thanks Amin! I haven't changed the height of the Message component, do the current Message components in stream-chat-react have dynamic heights?

**Comment by jaapbakker88** [comment]
@tigershen23 the messages do have dynamic heights. think of images and attachments. the browser doesnt know the height when they're being rendered. There's way's to fix this but still you won't have the best performance still. 

First version we're going to ship will come with a default message component with attachments/reactions disabled. Then slowly we'll try to add support for more features.

**Comment by tigershen23** [comment]
Hmm ok thanks for the context! Looking forward to trying it out.

**Comment by tigershen23** [comment]
Hey folks, is there any thought to being able to configure (or setting a lower number) for the number of messages that are loaded initially into memory (and unloading them as they get superceded by lots of old messages)? In my usage I've felt that this number is currently quite high and it seems like a quick performance win to be able to simply load fewer messages and components into memory. Can you provide some context behind the current number of "active" messages and maybe some color around what it would take to be able to configure this?

**Comment by mahboubii** [comment]
> Hey folks, is there any thought to being able to configure (or setting a lower number) for the number of messages that are loaded initially into memory (and unloading them as they get superceded by lots of old messages)? In my usage I've felt that this number is currently quite high and it seems like a quick performance win to be able to simply load fewer messages and components into memory. Can you provide some context behind the current number of "active" messages and maybe some color around what it would take to be able to configure this?

@tigershen23 We have discussed this internally and we might add a prop for it later, but a quick fix to this problem (if it fits your use case) is to copy the MessageList component and slice the messages array when it gets too big for you.

**Comment by tigershen23** [comment]
Got it, thanks @mahboubii !

File: src/styles/MessageNotification.scss [hunk 1]
```diff
@@ -8,6 +8,6 @@
   border-radius: 50px;
   padding: 4px 10px;
   font-size: 12px;
-  bottom: -10px;
+  bottom: 10px;
```
[the remark this answers is not in the document]
Comment (mahboubii) [reply src/styles/MessageNotification.scss:11]: yea manually checked with MessageList and it looks fine

File: src/components/MessageList/VirtualizedMessageList.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+import { EventComponent } from '../EventComponent';
+import { LoadingIndicator as DefaultLoadingIndicator } from '../Loading';
+import { EmptyStateIndicator as DefaultEmptyStateIndicator } from '../EmptyStateIndicator';
+
+/**
+ * VirtualizedMessageList - This component renders a list of messages in a virtual list. Its a consumer of [Channel Context](https://getstream.github.io/stream-chat-react/#channel)
+ * It is pretty fast for rendering thousands of messages but it needs its Message componet to have fixed height
+ * @example ../../docs/VirtualizedMessageList.md
+ */
+const VirtualizedMessageList = ({
```
Comment (vishalnarkhede) [inline src/components/MessageList/VirtualizedMessageList.js:23]: @mahboubii typescript?

Comment (mahboubii) [reply src/components/MessageList/VirtualizedMessageList.js:23]: to soon, will add typescript and some tests once we have some feedbacks

**Comment by tigershen23** [comment]
Hiya -- any update on when message actions (specifically message delete) will be added here? That's the last blocker before we can get rolling with testing this on our end

**Comment by tigershen23** [comment]
Tested this today and for the most part it went really well! Seems like we could use a null check here https://sentry.io/share/issue/6b4da5872bed49b8b700c0510b414c6e/

**Comment by tigershen23** [comment]
Fastest response ever, thanks @mahboubii  =) bump on https://github.com/GetStream/stream-chat-react/pull/487#issuecomment-688089557 and also when you're planning on 2.3.0-beta.2, whenever you get a chance

**Comment by tigershen23** [comment]
Best service ever; next time I won't even bug ya :)

**Comment by mahboubii** [comment]
> Fastest response ever, thanks @mahboubii =) bump on [#487 (comment)](https://github.com/GetStream/stream-chat-react/pull/487#issuecomment-688089557) and also when you're planning on 2.3.0-beta.2, whenever you get a chance
thanks for reporting this issue, `beta.2` published to npm 
https://www.npmjs.com/package/stream-chat-react/v/2.3.0-beta.2

**Comment by mahboubii** [comment]
> Hiya -- any update on when message actions (specifically message delete) will be added here? That's the last blocker before we can get rolling with testing this on our end

we still have to evaluate this one to see how it can be added properly, but we'll add it by next week 🤞

**Comment by tigershen23** [comment]
Sweet, thanks! I might take some time to look into it as well, since we do want to add some custom actions. Looking forward to your updates.

**Comment by mahboubii** [comment]
> Sweet, thanks! I might take some time to look into it as well, since we do want to add some custom actions. Looking forward to your updates.

@tigershen23 just released the `beta.3` which supports deleting a message, let us know if you had any issue with it.

**Comment by tigershen23** [comment]
@mahboubii amazing!! Will let you know any questions or feedback

**Review by jaapbakker88 — APPROVED** [review]
LGTM

**Pull request merged**